src/metabase/search/filter.clj

```diff
@@ -135,6 +130,38 @@
     (sql.helpers/where query [:= (search.config/column-with-model-alias model :authority_level) "official"])
     (sql.helpers/where query false-clause)))
 
+;; Created at filters
+
+(defn- default-created-at-filter-clause
+  [model created-at]
+  (let [{:keys [start end]} (try
+                             (params.dates/date-string->range created-at)
+                             (catch Exception _e
+                               (throw (ex-info (tru "Failed to parse created at param: {0}" created-at) {:status-code 400}))))
+        start               (when start (u.date/parse start))
+        end                 (when end (u.date/parse end))
+        created-at-col      (search.config/column-with-model-alias model :created_at)
+        created-at-col      (if (some #(instance? LocalDate %) [start end])
+                              [:cast created-at-col :date]
+                              created-at-col)]
+    (cond
+     (= start end)
+     [:= created-at-col start]
+
+     (nil? start)
+     [:< created-at-col end]
+
+     (nil? end)
+     [:> created-at-col start]
+
+     :else
+     [:and [:>= created-at-col start] [:< created-at-col end]])))
```

Comment:
This seems inconsistent with how date-string ranges with a start and end are treated in query parameters.

In SQL, a range like `"2016-07-01~2016-08-01"` is substituted with `<date-column> BETWEEN ? AND ?`. Here's an example test showing this behaviour:
https://github.com/metabase/metabase/blob/a2336e64c0105d61225593f500b1b58b06481940/test/metabase/driver/sql/parameters/substitute_test.clj#L561-L564

I think we should make search filters consistent with filters in query parameters, so for ranges like `"2016-07-01~2016-08-01"` we should include both the start and the end of the range, using `:between`.

Reply:
I think we'll want to change the behavior of that filter in query parameter to this version. see this comment: https://github.com/metabase/metabase/blob/830b4d9ee133da94e8d5186b12dd465e063acc3d/src/metabase/driver/sql/parameters/substitution.clj#L184

Reply:
There's something else going on here because I think this FIXME refers to a quirk of the internal implementation for date range substitution. I tried using a date range in a dashboard and it doesn't match your current implementation.

I added the Orders table to a dashboard, with a date parameter linked to created_at. Then I selected a date range for the parameter like `"2023-08-01~2023-08-08"`. 

The date picker clearly shows that `2023-08-08` should be included:
![image](https://github.com/metabase/metabase/assets/39073188/be7c5840-ceb2-4b24-b101-7a07a387a89b)

The query and parameters on the API request are below:
![image](https://github.com/metabase/metabase/assets/39073188/d949a490-df2d-4a73-b1f0-3a0b1ef2a153)

And the compiled query includes the where clause:
```
WHERE (\"PUBLIC\".\"ORDERS\".\"CREATED_AT\" >= timestamp with time zone '2023-08-01 00:00:00.000Z') 
  AND (\"PUBLIC\".\"ORDERS\".\"CREATED_AT\" < timestamp with time zone '2023-08-09 00:00:00.000Z')
```

So actually `"<start>~<end>"` is actually being converted to `<start> >= x > <end>`.

I think [this test demonstrates](https://github.com/metabase/metabase/pull/32836/files#diff-bbba19769b3cc1ed2910c6aba124a9881c8fb0c64d1a42976589fa315afdcff5R111-R112) this is a bug more clearly. If the date is `2023-05-04` I would expect "past3days" it should include yesterday but not today, so it should translate to `2023-05-01 >= x < 2023-05-04`, but instead in the test it translates to `2023-05-01 >= x < 2023-05-03`, which is only two full days.

Reply:
good thinking on the UI. I don't have a strong opinion regarding whether date range should be end-inclusive. I found that comment and I think that's what we want so that explains my current implementation.

Doing a quick search and it looks like "most" ppl do [A, B) for date range ([StackOverflow](https://stackoverflow.com/questions/9795391/is-there-a-standard-for-inclusive-exclusive-ends-of-time-intervals))

Regarding the `past3days` example, I think that's intended because we have a syntax to "include current". for your example, `past3days~` will translate to `2023-05-01 >= x < 2023-05-04`.

Reply:
> I don't have a strong opinion regarding whether date range should be end-inclusive.

Sorry, I should have clearer - in my last comment I should have conceded that the date range should be end-exclusive. However, I'm making the point the end looks like it should be extended by a unit, because it's off by one unit currently. I don't know how this is implemented with date range filters in dashboards, but it's clearly done differently to your current implementation given the example demonstration.

> Regarding the past3days example, I think that's intended because we have a syntax to "include current". for your example, past3days~ will translate to 2023-05-01 >= x < 2023-05-04.

That explanation doesn't make sense because "include current" version "past3days~" if today was '2023-05-04` should translate to `2023-05-01 >= x < 2023-05-05`, not `2023-05-01 >= x < 2023-05-04` which doesn't include today.

Reply:
You're right, that's the correct behavior to shift `end` by one unit.
I've modified to accommodate this.

FYI: For the QP this is done in `optimize-temporal-filters` middleware.

Reply:
Unresolving this because I had an idea. `date-string->range` can be changed to always return a range that is exclusive of the end and inclusive of the start, making this all a lot simpler. Apart from these search filters and [this usage in the mongo driver](https://github.com/metabase/metabase/blob/8c571f5d2a149b20b34ad28687951aa2aa819e91/modules/drivers/mongo/src/metabase/driver/mongo/parameters.clj#L81), native query parameters is the only other usage of `date-string->range`. If we fix SQL native query parameters to compile to `start <= x < end` instead of `x between start and end` then we can fix this FIXME (below), and switch the `:range` functions in the date decoders to always return a range that is exclusive of the end. Doing it this way wouldn't touch the QP and it would make this implementation simpler because you wouldn't need to pass around this unit stuff. You can just change `metabase.driver.common.parameters.dates/comparison-range` to call `metabase.util.date-2/comparison-range` with the `{:end :inclusive}` option

https://github.com/metabase/metabase/blob/8c571f5d2a149b20b34ad28687951aa2aa819e91/src/metabase/driver/sql/parameters/substitution.clj#L186-L192

Reply:
I'm not sure how is this simpler, we still need to pass the `:unit` around in `parameters.dates` so `adjust-inclusive-range-if-needed` could adjust the time correctly, rather than assuming everything has `day` unit like before.

Also, having both options as either `true` or `false` is more predictable if we think of that namespace as a library. 
How we use it is up to the call site, we shouldn't include the business logic there. But this is not the main point, I just don't see how much simpler that gives us with respect to the current change.

Reply:
does that mean we'll remove the `inclusive-start and inclusive-end` options of `date-string->range` too?

Reply:
Yes, we can

Reply:
I'll do this in another PR